java/src/Main.java

```diff
@@ -0,0 +1,155 @@
+            return "(%s %s %s)".formatted(left, operator, right);
+        }
+    }
+
+    static class Select {
+        public final Collection<String> columns;
+        Select(String... columns) {
+            this.columns = Arrays.asList(columns);
+        }
+        public String toString() {
+            return String.join(", ", columns);
+        }
+    }
+
+    static class Field {
+        public String description;
+        public String mode;
+        public String name;
+        public String type;
+        public Field[] fields;
+    }
+
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    static class Schema {
+        public Field[] fields;
+    }
+
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    static class SchemaRoot {
+        public Schema schema;
+    }
+
+    static Map<String, String> makeUnnestDictionary() throws IOException {
+        var schemaFile = new File("queryt/schema.json");
+        var root = new ObjectMapper().readValue(schemaFile, SchemaRoot.class);
+        var unnestDictionary = new HashMap<String, String>();
+        for (Field v : root.schema.fields) {
+            if (v.type.equals("RECORD")) {
+                for (Field field : v.fields) {
+                    unnestDictionary.put(field.name, v.name);
+                }
+            }
+        }
+        return unnestDictionary;
+    }
+
+    static class Query {
+        public final String table;
+        public final Select select;
+        public final Where where;
+
+        public Query(String table, Select select, Where where) {
+            this.table = table;
+            this.select = select;
+            this.where = where;
+        }
+
+        String translate(Map<String, String> unnestDict) {
+            var fromClause =
+                    Stream.concat(Stream.of(table),
+                            Stream.concat(select.columns.stream(), where.columns().stream())
+                                    .filter(unnestDict::containsKey)
+                                    .map(unnestDict::get)
+                                    .distinct()
+                                    .map("unnest(%s)"::formatted))
+                                    .collect(Collectors.joining(", "));
+
+            return "SELECT %s FROM %s WHERE %s".formatted(select, fromClause, where);
+        }
+    }
+
+    public static void main(String[] args) throws IOException {
+
+        var unnestDict = makeUnnestDictionary();
+
+        var w1 = new Where(new Column("age_at_index"), ">=", new Value(50));
+        var w2 = new Where(new Column("project_id"), "like", new Value("TCGA%"));
+        var w3 = new Where(new Column("figo_stage"), "=", new Value("Stage IIIC"));
+
+        var w4 = new Where(w1, "and", w2);
+        var w5 = new Where(w4, "and", w3);
+
+        var s = new Select("case_id", "age_at_index", "gender", "race", "project_id", "figo_stage");
+
+        var q = new Query("gdc-bq-sample.gdc_metadata.r24_clinical", s, w5);
```

Comment:
I did diverge from the python code here because I wanted to create an object I could serialize to Json that would stand alone as a query.

My thinking is that either the schema would be "baked in" to the database side and wouldn't be part of the query itself, so I split it out.